raft.go

```diff
@@ -812,15 +921,16 @@ func (r *Raft) appendEntries(rpc RPC, a *AppendEntriesRequest) {
 
 	// Increase the term if we see a newer one, also transition to follower
 	// if we ever get an appendEntries call
-	if a.Term > r.getCurrentTerm() || r.getState() != Follower {
-		// Ensure transition to follower
-		r.setState(Follower)
-		r.setCurrentTerm(a.Term)
+	if a.Term > r.getCurrentTerm() {
+		r.updateTerm(a.Term)
 		resp.Term = a.Term
 	}
-
 	// Save the current leader
+	r.stepDown()
 	r.setLeader(ServerAddress(r.trans.DecodePeer(a.Leader)))
+	defer r.setLastContact()
+	// TODO: moving setLastContact up here is the right thing to do, but it poses a compatibility problem.
+	// It may be better to send heartbeats of (index 0, term 0) to old servers or something.
```

Comment:
there's a big comment in the latest code explaining why it uses the matchIndex instead now